### PR TITLE
Log to app insights when using console methods

### LINKED ISSUE #1715: Override console function with call to application insights in cloud enviornemtns
### Description
To make the capturing of fronted logging easy, consistent, and complete we should look at overriding the existing console calls to make a call to Azure insights

This would give us the same logging in dev and local as we get in front, which makes it easier to implement front end code. This would improve our telemetry in deployed environments as well. 

### Proposed solution 
see https://stackoverflow.com/a/30197438

### PR DESCRIPTION
**Pull request by nickscialli-usds**
## Related Issue or Background Info

- Closes #1715

## Changes Proposed

- Refactor telemetry to use typescript and hooks
- Add a `withInsights` function that takes the `console` as a argument and adds telemetry to the log, info, warn, and error methods.

## Additional Information

- One potential concern with this approach: 
    - ~~If app insights itself calls a console method it could resulting in an infinite loop. I tried to head off the one known instance of this if the argument passed to the logger matches the regex `/failed to send telemetry/i`, but this seems hacky and will fail for any other potential instances where `appInsights` might end up calling an affected console method.~~ I believe this is no longer a concern after setting `loggingLevelConsole` to `0`.

## Screenshots / Demos

- `console.error` reported to the test environment can be [viewed in App Insights here](https://portal.azure.com/#blade/AppInsightsExtension/DetailsV2Blade/ComponentId/%7B%22SubscriptionId%22%3A%227d1e3999-6577-4cd5-b296-f518e5c8e677%22%2C%22ResourceGroup%22%3A%22prime-simple-report-test%22%2C%22Name%22%3A%22prime-simple-report-test-insights%22%2C%22LinkedApplicationType%22%3A0%2C%22ResourceId%22%3A%22%252Fsubscriptions%252F7d1e3999-6577-4cd5-b296-f518e5c8e677%252FresourceGroups%252Fprime-simple-report-test%252Fproviders%252FMicrosoft.Insights%252Fcomponents%252Fprime-simple-report-test-insights%22%2C%22ResourceType%22%3A%22microsoft.insights%252Fcomponents%22%2C%22IsAzureFirst%22%3Afalse%7D/DataModel/%7B%22eventId%22%3A%227ae82870-db4a-11eb-9415-d7a8e5f10c08%22%2C%22timestamp%22%3A%222021-07-02T15%3A30%3A50.126Z%22%2C%22cacheId%22%3A%226d41a512-4606-447c-a47d-ab36c0ecb5fe%22%2C%22eventTable%22%3A%22exceptions%22%2C%22timeContext%22%3A%7B%22durationMs%22%3A3600000%2C%22endTime%22%3A%222021-07-02T15%3A34%3A55.970Z%22%7D%7D)
- `consolee.log` reported as a custom event can be [viewed in App Insights here](https://portal.azure.com/#blade/AppInsightsExtension/DetailsV2Blade/ComponentId/%7B%22SubscriptionId%22%3A%227d1e3999-6577-4cd5-b296-f518e5c8e677%22%2C%22ResourceGroup%22%3A%22prime-simple-report-test%22%2C%22Name%22%3A%22prime-simple-report-test-insights%22%2C%22LinkedApplicationType%22%3A0%2C%22ResourceId%22%3A%22%252Fsubscriptions%252F7d1e3999-6577-4cd5-b296-f518e5c8e677%252FresourceGroups%252Fprime-simple-report-test%252Fproviders%252FMicrosoft.Insights%252Fcomponents%252Fprime-simple-report-test-insights%22%2C%22ResourceType%22%3A%22microsoft.insights%252Fcomponents%22%2C%22IsAzureFirst%22%3Afalse%7D/DataModel/%7B%22eventId%22%3A%22811d18eb-db4b-11eb-9f98-e1cde2fe9e25%22%2C%22timestamp%22%3A%222021-07-02T15%3A38%3A09.980Z%22%2C%22cacheId%22%3A%22608b148f-04c0-46e9-9170-91759e88f57e%22%2C%22eventTable%22%3A%22customEvents%22%2C%22timeContext%22%3A%7B%22durationMs%22%3A1800000%2C%22endTime%22%3A%222021-07-02T15%3A45%3A52.554Z%22%7D%7D)
### UI
- [n/a] Any changes to the UI/UX are approved by design 
- [n/a] Any new or updated content (e.g. error messages) are approved by design 

### Testing
- [x] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [n/a] Database changes are submitted as a separate PR
  - [n/a] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [n/a] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
        (including re-granting permission to the no-PHI user if need be)
  - [n/a] Liquibase rollback has been tested locally using `./gradlew liquibaseRollbackSQL` or `liquibaseRollback`
- [n/a] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [n/a] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [n/a] Any dependencies introduced have been vetted and discussed

## Cloud
- [n/a] DevOps team has been notified if PR requires ops support
- [x] If there are changes that cannot be tested locally, this has been deployed to our Azure `test` environment for verification

### CODE REVIEW

File: frontend/src/app/TelemetryService.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    console[method] = (...data: any[]) => {
+      originalConsole[method](data);
+
+      // Prevent telemetry failure infinite loop
+      if (telemetryFailure.test(data[0])) {
+        return;
+      }
+
+      if (method === "error" || method === "warn") {
```
Comment (nickscialli-usds) [inline frontend/src/app/TelemetryService.ts:74]: Will report errors and warnings as exceptions in app insights and the rest as custom events. Happy to change this if something else makes more sense.

**Comment by nullflux** [comment]
> If app insights itself calls a console method it could resulting in an infinite loop. I tried to head off the one known instance of this if the argument passed to the logger matches the regex /failed to send telemetry/i, but this seems hacky and will fail for any other potential instances where appInsights might end up calling an affected console method.

This is a real problem. I dug a bit into the AppInsights source, and to my dismay there's a fair amount of `console.warn`ing, both through their `warnToConsole` and `throwInternal` methods. Regrettably, they're not prefixed in any way nor do they share any easily greppable patterning. Here's a sampling, beginning of course w/ the one you mention:

```
_self.diagLog().throwInternal(LoggingSeverity.WARNING, _InternalMessageId.OnError, "Failed to send telemetry.", { message: message });
itemCtx.diagLog().warnToConsole("Telemetry processor check returns false");
logger.warnToConsole("Invalid input for dependency data");
logger.warnToConsole("Two extensions have same priority #" + extPriority + " - " + extPriorities[extPriority] + ", " + identifier);
this._logger.warnToConsole("Auto track page visit time failed, metric will not be collected: " + Util.dump(e));
this._logger.warnToConsole("Call to restart failed: " + Util.dump(e));
this._logger.warnToConsole("Call to start failed: " + Util.dump(e));
this._logger.warnToConsole("Stop page visit timer failed: " + Util.dump(e));
itemCtx.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.InvalidEvent, "Cannot send telemetry without baseData and baseType");
itemCtx.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.PluginException, "Plugin [" + plugin.identifier + "] failed during processTelemetry - " + error);
itemCtx.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.SenderNotInitialized, "Sender was not initialized")
this.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.StartTrackFailed, "startTrackPage failed, page view may not be collected: " + Util.getExceptionName(e), { exception: Util.dump(e) });
this.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.StopTrackEventFailed, "stopTrackEvent failed, event will not be collected: " + Util.getExceptionName(e), { exception: Util.dump(e) });
this.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.StopTrackFailed, "stopTrackPage failed, page view will not be collected: " + Util.getExceptionName(e), { exception: Util.dump(e) });
_self.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.NotificationException, "send request notification failed: " + Util.getExceptionName(e), { exception: Util.dump(e) });
_self.diagLog().throwInternal(LoggingSeverity.CRITICAL, _InternalMessageId.TransmissionFailed, "Telemetry transmission failed, some telemetry will be lost: " + Util.getExceptionName(e), { exception: Util.dump(e) });
```

I think we should take this up with @timothybest-usds on Tuesday, and consider instead using a `logger` util and instead just entirely avoiding `console` by convention.

**Comment by nickscialli-usds** [comment]
@nullflux yes, I considered that as well re: a custom logger. That solves the problem but of course it only means we get insights in places we call our custom logger and not if we're hitting an error in a dependency, for example

**Comment by nullflux** [comment]
That's fair, but I would expect that the type of critical dependency failures that we'd need to know about would throw exceptions rather than just complaining to the console – and we should already hear about those via AppInsights because of the top-level error boundary

File: frontend/src/app/TelemetryService.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  ApplicationInsights,
+  SeverityLevel,
+} from "@microsoft/applicationinsights-web";
+import { ReactPlugin } from "@microsoft/applicationinsights-react-js";
+import { useHistory } from "react-router";
+
+let reactPlugin: ReactPlugin | null = null;
+let appInsights: ApplicationInsights | null = null;
+
+const createTelemetryService = () => {
```
Comment (nullflux) [inline frontend/src/app/TelemetryService.ts:11]: **praise:** Thanks for converting this to typescript!

**Comment by nickscialli-usds** [comment]
@nullflux whoops, I think the solution is in the app insights config object:

>loggingLevelConsole: Console logging level. All logs with a severity level higher than the configured level will be printed to console. Otherwise they are suppressed. ie Level 2 will print both CRITICAL and WARNING logs to console, level 1 prints only CRITICAL.
> 
> Note: Logs sent as telemetry to instrumentation key will also be logged to console if their severity meets the configured loggingConsoleLevel
> 
> 0: ALL console logging off 1: logs to console: severity >= CRITICAL 2: logs to console: severity >= WARNING

File: frontend/src/app/TelemetryService.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      return;
+    }
+
+    reactPlugin = new ReactPlugin();
+
+    appInsights = new ApplicationInsights({
+      config: {
+        instrumentationKey,
+        extensions: [reactPlugin],
+        loggingLevelConsole: process.env.NODE_ENV === "development" ? 2 : 0,
```
Comment (nickscialli-usds) [inline frontend/src/app/TelemetryService.ts:26]: I was able to confirm that setting `loggingLevelConsole` to 0 stops all console logging from AppInsights itself, which should prevent infinite loop concerns.

File: frontend/src/app/TelemetryService.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      return;
+    }
+
+    reactPlugin = new ReactPlugin();
+
+    appInsights = new ApplicationInsights({
+      config: {
+        instrumentationKey,
+        extensions: [reactPlugin],
+        loggingLevelConsole: 2,
```
Comment (nullflux) [inline frontend/src/app/TelemetryService.ts]: **thought:** Related to the above conversation about infinite loops: another option for that could be passing `loggingLevelConsole: 0` here to the `ApplicationInsights` config at initialization? I'm a little fuzzy abt the AppInsights logic here, but I know that at least will cover `throwInternal`, if not explicitly `warnToConsole`

File: frontend/src/app/TelemetryService.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  error: SeverityLevel.Error,
+  info: SeverityLevel.Information,
+} as const;
+
+const telemetryFailure = /failed to send telemetry/i;
+
+export function withInsights(console: Console) {
+  const originalConsole = { ...console };
+
+  Object.entries(logSeverityMap).forEach((el) => {
```
Comment (nullflux) [inline frontend/src/app/TelemetryService.ts:58]: **praise:** Thanks for making this so DRY in the first place!

File: frontend/src/app/TelemetryService.ts (hunk 3 above)
Comment (nullflux) [reply frontend/src/app/TelemetryService.ts:26]: Excellent! I just submitted a similar comment in my review; so we can just `resolve` that convo.

And with that, I say this whole situation is ✅

**Review by nullflux — APPROVED** [review]

File: frontend/src/app/TelemetryService.ts (hunk 4 above)
Comment (nullflux) [reply frontend/src/app/TelemetryService.ts]: 🎉

**Pull request merged**